core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,449 @@
+    private static final int MAX_TOMBSTONES = 10000;
+
+    // Setting for the expiration window of tombstone entries.  Any tombstone entries
+    // that have been in the list longer than EXPIRATION_WINDOW will be purged on the
+    // next cluster update.
+    static final long EXPIRATION_WINDOW = 1000 * 60 * 60 * 24 * 14; // 14 days
+
+    // holds the index tombstones
+    private final Deque<Tombstone> tombstones;
+
+    public IndexGraveyard() {
+        tombstones = new ArrayDeque<>();
+    }
+
+    // copy constructor
+    public IndexGraveyard(final IndexGraveyard that) {
+        this.tombstones = new ArrayDeque<>(that.tombstones);
+    }
+
+    public IndexGraveyard(final StreamInput in) throws IOException {
+        final int queueSize = in.readVInt();
+        tombstones = new ArrayDeque<>(queueSize);
+        for (int i = 0; i < queueSize; i++) {
+            tombstones.add(new Tombstone(in));
+        }
+    }
+
+    // only used by fromXContent
+    private IndexGraveyard(final List<Tombstone> list) {
+        this.tombstones = new ArrayDeque<>(list);
+    }
+
+    // only used by IndexGraveyardDiff
+    private IndexGraveyard(final Deque<Tombstone> queue) {
+        this.tombstones = new ArrayDeque<>(queue);
+    }
+
+    /**
+     * Do they have the same index tombstones (deletions)?
+     */
+    public boolean isEquals(IndexGraveyard that) {
+        // array deque doesn't implement equals
+        return Objects.equals(new ArrayList<>(tombstones), new ArrayList<>(that.tombstones));
+    }
+
+    /**
+     * Add a deleted index to the list of tombstones in the cluster state.  Returns the
+     * purged tombstone if the size of the list grew too big, else return none.
+     */
+    public Optional<Tombstone> addTombstone(final Index index) {
+        return addTombstone(index, System.currentTimeMillis(), System.nanoTime());
+    }
+
+    /**
+     * Add a deleted index to the list of tombstones in the cluster state, along with
+     * a specified deletion date and nanosecond date.  Returns the purged tombstone
+     * if the size of the list grew too big, else return none.
+     */
+    Optional<Tombstone> addTombstone(final Index index, final long deletionDate, final long deletionDateNS) {
+        final Optional<Tombstone> purged;
+        if (tombstones.size() == MAX_TOMBSTONES) {
+            // purge the oldest entry
+            purged = Optional.of(tombstones.removeFirst());
+        } else {
+            purged = Optional.empty();
+        }
+        tombstones.addLast(new Tombstone(index, deletionDate, deletionDateNS));
+        return purged;
+    }
+
+    /**
+     * Get a copy of the current index tombstones.
+     */
+    List<Tombstone> getTombstones() {
+        return new ArrayList<>(tombstones);
+    }
+
+    /**
+     * Purge tombstone entries by date, if they have passed a certain expiration window.
+     * Returns the number of entries that were purged.
+     */
+    public int purge() {
```

Comment:
this method can the max size and min expiration window as parameters, which will make it easier to test and have those be settings,